### DISCUSSION TITLE
Performance/Memory impacts of Relationship Mappings

### DISCUSSION THREAD

**Discussion by vkomarov-r7** [sqlalchemy/sqlalchemy#7822] [Usage Questions] [answered]
This question originated on gitter surrounding the performance/memory impacts of relationship mapping operations in SQLAlchemy.

Context: We have a rather large Python application with 421 tables. I spent some time doing some profiling with the goal of reducing baseline memory utilization and improving startup time. After some sleuthing, I discovered that the `configure_mappers()` step (which in our case appears to be internally called on first use of a model object) takes about 2.5 seconds to run, and takes up about 110MB (!) of memory.

![image](https://user-images.githubusercontent.com/65631117/158617810-4ccd424b-2834-47f3-b038-f9e7f7a21ec1.png)

A big part of this 2.5 seconds appears to be part of `relationships.py::do_init()`, which is called 2851 times, taking 2.449 seconds. Internally, the following calls make up the majority of the time:
* _setup_join_conditions: 1.52s
* _process_dependent_arguments: 0.47s
* _generate_backref: 0.41s
* do_init (self): 0.29s

I performed some `objgraph` analysis to figure out what objects are being created as part of the configure mappers step, and this is what I got:

```
dict                         191619   +127399
set                           48739    +26856
AnnotatedColumn               26834    +26799
BinaryExpression              25381    +24860
Event                         23241    +23240
tuple                         40637    +21969
weakref                       30637    +19293
Comparator                    27038    +16480
function                      50997    +12054
list                          36279    +11937
method                        18625     +9533
deque                         10612     +8546
_ListenerCollection           10451     +8538
cell                          14348     +7723
AttributeEventsDispatch        7086     +7085
BindParameter                  7501     +6367
_anonymous_label               7501     +6367
ColumnLoader                   4239     +4239
ScalarAttributeImpl            4239     +4239
BooleanClauseList              3002     +3002
JoinCondition                  2846     +2846
LazyLoader                     2846     +2846
_GetColumns                    2470     +2470
ManyToOneDP                    2046     +2046
DetectKeySwitch                2046     +2046
CollectionAttributeImpl        1987     +1987
ScalarObjectAttributeImpl       858      +858
InstrumentedAttribute          7085      +376
RelationshipProperty           2846      +376
CascadeOptions                 2846      +376
JoinedLoader                    350      +350
OneToManyDP                     258      +258
ManyToManyDP                    194      +194
builtin_function_or_method     2502       +47
WeakSequence                    379       +34
ImmutableProperties              31       +31
_EmptyListener                  151        +8
OrderedDict                    2568        +7
WeakKeyDictionary               124        +4
DynaLoader                        1        +1
DynamicAttributeImpl              1        +1
```

Digging a layer deeper, I took a look at the number of BinaryExpressions created (which seems to represent join criteria), and found that there are many clones seemingly being created for each item, for example:

```
BinaryExpression<left=WebApps.resource_id, right=ServerlessFunctions.web_app_resource_id, operator=<built-in function eq>>
BinaryExpression<left=WebApps.resource_id, right=ServerlessFunctions.web_app_resource_id, operator=<built-in function eq>>
BinaryExpression<left=WebApps.resource_id, right=ServerlessFunctions.web_app_resource_id, operator=<built-in function eq>>
BinaryExpression<left=WebApps.resource_id, right=ServerlessFunctions.web_app_resource_id, operator=<built-in function eq>>
BinaryExpression<left=WebApps.resource_id, right=ServerlessFunctions.web_app_resource_id, operator=<built-in function eq>>
BinaryExpression<left=WebApps.resource_id, right=ServerlessFunctions.web_app_resource_id, operator=<built-in function eq>>
BinaryExpression<left=WebApps.resource_id, right=ServerlessFunctions.web_app_resource_id, operator=<built-in function eq>>
BinaryExpression<left=WebApps.resource_id, right=ServerlessFunctions.web_app_resource_id, operator=<built-in function eq>>
BinaryExpression<left=WebApps.resource_id, right=ServerlessFunctions.web_app_resource_id, operator=<built-in function eq>>
BinaryExpression<left=WebApps.resource_id, right=ServerlessFunctions.web_app_resource_id, operator=<built-in function eq>>
BinaryExpression<left=WebApps.resource_id, right=ServerlessFunctions.web_app_resource_id, operator=<built-in function eq>>
BinaryExpression<left=WebApps.resource_id, right=ServerlessFunctions.web_app_resource_id, operator=<built-in function eq>>
BinaryExpression<left=WebApps.resource_id, right=ServerlessFunctions.web_app_resource_id, operator=<built-in function eq>>
```

This relationship is declared like this (other details omitted):

```python
class WebApp(Base):
    __tablename__ = 'WebApps'

    id = Column(Integer, autoincrement=True)
    resource_id = Column(String, primary_key=True)

    serverless_functions = relationship(
        'ServerlessFunction', uselist=True, cascade='',
        backref=backref('webapp', cascade='', uselist=False)
    )

# in another module:
class ServerlessFunction(Base):
    __tablename__ = 'ServerlessFunctions'
    id = Column(Integer, autoincrement=True)
    resource_id = Column(String, primary_key=True)
```

Here's what the backref chain looks like for the above BinaryExpressions, seems like a lot of clones:

![backrefs](https://user-images.githubusercontent.com/65631117/158623496-7d31c6ea-82e1-458a-934c-9cd496ed6be8.png)

Not sure if there's something we're doing wrong (it seems pretty straightforward) or if there's something going on with SQLAlchemy internally.

## Environment

* Python 3.8.5
* SQLAlchemy 1.3.24 (note: this is an old version)

**Comment by zzzeek** [comment, marked as answer]
hi -

nope, you're not doing anything wrong.      When relationships set up, they take the given "primaryjoin" (or figure one out) and apply many passes of mutation to it, where you would see the difference mostly in a dictionary on the objects called "_annotations".

The cloning process keeps a reference to the thing it was cloned from, which is in many cases pretty important.

However, what you appear to be observing is that if we for example take a binary expression and clone it 9 times, where we really only need the first copy and the last, we still hold onto the 7 intermediary versions of it due to the behavior of "is_clone_of".

To improve upon this would require to alter the way in which these clones are created, such that in places we are cloning something and know that we want to throw away the intermediary version, we pass through some indication that "is_clone_of" should skip the intermediary and point to the previous item.     

This is certainly a doable improvement and I will add an issue for 2.0.  Unfortunately this doesnt provide any solution for the 1.3 or 1.4 series.

**Reply by zzzeek** [reply]
some poking around suggests it might be pretty easy as the is_clone_of is not consulted too often.      1.4 is still very late in the game though.  

in your 1.3.x release you might try this patch:

```diff
diff --git a/lib/sqlalchemy/sql/elements.py b/lib/sqlalchemy/sql/elements.py
index c69a9b18c..1978ddfb9 100644
--- a/lib/sqlalchemy/sql/elements.py
+++ b/lib/sqlalchemy/sql/elements.py
@@ -217,7 +217,8 @@ class ClauseElement(Visitable):
         # process leaves around a lot of remnants of the previous clause
         # typically in the form of column expressions still attached to the
         # old table.
-        c._is_clone_of = self
+        root_clone_of = self._is_clone_of
+        c._is_clone_of = self if root_clone_of is None else root_clone_of
 
         return c
```

that seems to reduce the number of BinaryExpression objects for a plain relationship from about nine to five.  taking out is_clone_of entirely leaves it at four.    Some initial testing suggests reducing clone_of at the creation site might work for all tests, having gone through everything yet.  removing clone_of completely fails some tests in the 2.0 series, which would be the same or worse in 1.3.

**Reply by zzzeek** [reply]
the change seems relatively low risk so I merged to 1.4 as well, so this particular aspect of memory use should be improved in the next 1.4 release

**Reply by vkomarov-r7** [reply]
Just had a chance to apply it, and that definitely helps in a measurable way. The total number of BinaryExpression allocations went from 25381 down to 13402, which netted a 80MB decrease in total memory and reduced startup time by ~0.5 seconds.

Thanks for looking into this! I think there's still more threads to pull, as the `configure_mappers()` step still takes on the order of 3.28s for us. Here is a snakeviz visualization as well as the associated profile. Note that the data is just for the `configure_mappers()` call after (most) of our DB models have been declared as shown above. It also includes the above root clone changes.

![image](https://user-images.githubusercontent.com/65631117/158860373-fdecf5b4-390d-412f-8e13-27bae1aa4e7d.png)
[stats.prof.zip](https://github.com/sqlalchemy/sqlalchemy/files/8288527/stats.prof.zip)

**Reply by zzzeek** [reply]
yeah the ORM is oriented towards trying to optimize for runtime speed, not startup speed, so the mapping process has a big emphasis on setting up structures that can invoke common operations more quickly.   so that's what you're seeing there

**Reply by zzzeek** [reply]
I will tell you what we added to 1.4 that provides at least a strategy to reducing startup time, which is that if your application has many different groups of mappers that are used in different ways, you can configure them in groups using different registries.   that is, if you have just one declarative_base() right now, you'd make several to indicate different sub-groups of mappers.   In 1.4, the "configuration" step only runs within the scope of a single registry, provided the mappers in that registry dont refer to mappers in some other registry.

that way if you can organize multiple Base classes into a uni-directional hierarchy of mappers, the configuration step would not occur all at once on startup if you didnt use all 400 mappings at once.

**Reply by zzzeek** [reply]
welp #7903 is a regression from this "trim intermediary clone" change, in the area where i would expect a regression to happen if at all.  i will attempt to make the "trim intermediaries" logic conditional so that it can be maintained for that one particular critical area.
